File: src/worker.ts
import { parentPort } from 'worker_threads';
import * as mysql from './WebConfig/lib/mysql';
import { calculate, calculateSubExpr } from './common/Expr';
import { BotInfo, ExprArgs, NODE_TYPE, Node, NodeData, ORDER_STATUS, RateData, TelegramIdType, UNIT, WorkerData } from './common/Interface';
import Telegram from './common/telegram';
import * as util from './common/util';

let botChildren: Array<BotInfo> = [];
const telegram = new Telegram(undefined, undefined, false);
let botIDs: { [key: string]: number } = {};
function onCloseCandle(broker: string, symbol: string, timeframe: string, data: Array<RateData>) {
    for (const botInfo of botChildren) {
        try {
            const { botName, idTelegram, symbolList, timeframes, route } = botInfo;
            if (!timeframes.includes(timeframe) || !symbolList.includes(`${broker}:${symbol}`)) continue;

            // console.log("onCloseCandle", { symbol, timeframe });

            const visited: { [key: string]: boolean } = {};
            dfs_handleLogic(route, broker, symbol, timeframe, data, idTelegram, visited, botIDs[botName]);
        }
        catch (err) {
            console.error({ symbol, timeframe }, err);
        }
    }
}

function binarySearch(arr: Array<string>, target: string): boolean {
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] === target) {
            return true;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return false;
}


function dfs_handleLogic(node: Node, broker: string, symbol: string, timeframe: string, data: RateData[], idTelegram: TelegramIdType, visited: { [key: string]: boolean }, botID: number) {
    const { id, next } = node;
    const nodeData = node.data;

    if (visited[id] === true) return;
    visited[id] = true;
    if (handleLogic(nodeData, broker, symbol, timeframe, data, idTelegram, botID)) {
        for (const child of next) {
            dfs_handleLogic(child, broker, symbol, timeframe, data, idTelegram, visited, botID);
        }
    }
}

function handleLogic(nodeData: NodeData, broker: string, symbol: string, timeframe: string, data: RateData[], idTelegram: TelegramIdType, botID: number): boolean {
    if (nodeData.type === NODE_TYPE.START) return true;

    const exprArgs: ExprArgs = {
        broker,
        symbol,
        timeframe,
        data
    };

    if (nodeData.type === NODE_TYPE.EXPR) {
        if (!nodeData.value) return false;

        let expr = nodeData.value;
        expr = calculateSubExpr(expr, exprArgs);

        const result = calculate(expr, exprArgs);
        return Boolean(result);
    }

    if (nodeData.type === NODE_TYPE.TELEGRAM) {
        if (!nodeData.value) return false;

        let content: string = nodeData.value.trim();
        content = calculateSubExpr(content, exprArgs);

        const emoji: { [key: string]: string } = {
            'binance': '🥇🥇🥇',
            'bybit': '',
            'okx': '🏁🏁🏁',
            'binance_future': '🥇🥇🥇',
            'bybit_future': ''
        }
        const url: { [key: string]: string } = {
            'binance': `https://www.binance.com/en/trade/${symbol}?_from=markets&type=spot`,
            'bybit': `https://www.bybit.com/vi-VN/trade/spot/${symbol.replace('USDT', '')}/USDT`,
            'okx': `https://www.okx.com/vi/trade-spot/${symbol}`,
            'binance_future': `https://www.binance.com/en/futures/${symbol}?_from=markets`,
            'bybit_future': `https://www.bybit.com/trade/usdt/${symbol}`
        }

        let mess = emoji[broker];
        mess += `\n<a href="${url[broker]}"><b>${symbol}</b></a>`;
        mess += `\n${broker}`;
        mess += `\n${timeframe} ${data[0].timestring}`;
        mess += `\n${content}`;

        if (content === '<--->') mess = '--------------------';

        const ids = idTelegram.toString().split(',').map(item => item.trim()).filter(item => item !== '');
        for (const id of ids) {
            telegram.sendMessage(mess, id);
        }
        return true;
    }

    const node: NodeData = { ...nodeData };
    if (!adjustParam(node, exprArgs)) return false;

    if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT, NODE_TYPE.SELL_MARKET, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(node.type)) {
        const sql = `INSERT INTO Orders(symbol,broker,timeframe,orderType,volume,stop,entry,tp,sl,status,createdTime,expiredTime,botID) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?)`;
        const args = [
            symbol,
            broker,
            timeframe,
            node.type,
            node.volume,
            node.stop,
            node.entry,
            node.tp,
            node.sl,
            ORDER_STATUS.OPENED,
            util.nextTime(data[0].startTime, timeframe),
            node.expiredTime,
            botID
        ];
        console.log(`new order`, args);

        mysql.query(sql, args).catch(err => {
            console.error('mysql query error in worker thread', sql, args, err);
        });
        return true;
    }

    return false;
}

function adjustParam(data: NodeData, args: ExprArgs): boolean {
    //stop
    if ([NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT].includes(data.type)) {
        if (!data.stop) return false;
        let expr: string = data.stop;
        expr = calculateSubExpr(expr, args);
        if (data.unitStop === UNIT.PERCENT) expr = `close() * (100 + abs(${expr})) / 100`;
        data.stop = calculate(expr, args);
        if (data.stop === null) return false;
    }
    else if ([NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.stop) return false;
        let expr: string = data.stop;
        expr = calculateSubExpr(expr, args);
        if (data.unitStop === UNIT.PERCENT) expr = `close() * (100 - abs(${expr})) / 100`;
        data.stop = calculate(expr, args);
        if (data.stop === null) return false;
    }
    else {
        data.stop = undefined;
    }

    //entry
    if ([NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_LIMIT].includes(data.type)) {
        if (!data.entry) return false;
        let expr: string = data.entry;
        expr = calculateSubExpr(expr, args);
        if (data.unitEntry === UNIT.PERCENT) expr = `close() * (100 - abs(${expr})) / 100`;
        data.entry = calculate(expr, args);
        if (data.entry === null) return false;
    }
    else if ([NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.entry) return false;
        let expr: string = data.entry;
        expr = calculateSubExpr(expr, args);
        if (data.unitEntry === UNIT.PERCENT) expr = `close() * (100 + abs(${expr})) / 100`;
        data.entry = calculate(expr, args);
        if (data.entry === null) return false;
    }
    else if ([NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.SELL_STOP_MARKET].includes(data.type)) {
        data.entry = data.stop;
    }
    else if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.SELL_MARKET].includes(data.type)) {
        data.entry = args.data[0].close;
    }

    const closePrice: number = args.data[0].close;
    const entry: number = parseFloat(data.entry);
    const stop: number = parseFloat(data.stop);

    // match entry immediately
    if (data.type === NODE_TYPE.BUY_LIMIT && closePrice <= entry) {
        data.entry = closePrice;
    }
    else if (data.type === NODE_TYPE.BUY_STOP_LIMIT && closePrice <= entry && closePrice >= stop) {
        data.entry = closePrice;
    }
    else if (data.type === NODE_TYPE.SELL_LIMIT && closePrice >= entry) {
        data.entry = closePrice;
    }
    else if (data.type === NODE_TYPE.SELL_STOP_LIMIT && closePrice >= entry && closePrice <= stop) {
        data.entry = closePrice;
    }

    //sl
    if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT].includes(data.type)) {
        if (!data.sl) return false;
        let expr: string = data.sl;
        expr = calculateSubExpr(expr, args);
        if (data.unitSL === UNIT.PERCENT) expr = `(${data.entry}) * (100 - abs(${expr})) / 100`;
        data.sl = calculate(expr, args);
        if (data.sl === null) return false;
    }
    else if ([NODE_TYPE.SELL_MARKET, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.sl) return false;
        let expr: string = data.sl;
        expr = calculateSubExpr(expr, args);
        if (data.unitSL === UNIT.PERCENT) expr = `(${data.entry}) * (100 + abs(${expr})) / 100`;
        data.sl = calculate(expr, args);
        if (data.sl === null) return false;
    }
    else {
        data.sl = undefined;
    }

    //tp
    if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT].includes(data.type)) {
        if (!data.tp) return false;
        let expr: string = data.tp;
        expr = calculateSubExpr(expr, args);
        if (data.unitTP === UNIT.PERCENT) expr = `(${data.entry}) * (100 + abs(${expr})) / 100`;
        else if (data.unitTP === UNIT.RR) expr = `(${data.entry} + abs(${data.entry} - ${data.sl}) * abs(${expr}))`;
        data.tp = calculate(expr, args);
        if (data.tp === null) return false;
    }
    else if ([NODE_TYPE.SELL_MARKET, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.tp) return false;
        let expr: string = data.tp;
        expr = calculateSubExpr(expr, args);
        if (data.unitTP === UNIT.PERCENT) expr = `(${data.entry}) * (100 - abs(${expr})) / 100`;
        else if (data.unitTP === UNIT.RR) expr = `(${data.entry} - abs(${data.entry} - ${data.sl}) * abs(${expr}))`;
        data.tp = calculate(expr, args);
        if (data.tp === null) return false;
    }
    else {
        data.tp = undefined;
    }

    //volume
    if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT, NODE_TYPE.SELL_MARKET, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.volume) return false;
        let expr: string = data.volume;
        expr = calculateSubExpr(expr, args);
        if (data.unitVolume === UNIT.USD) expr = `(${expr}) / ${data.entry}`;
        data.volume = calculate(expr, args);
        if (data.volume === null) return false;
    }
    else {
        data.volume = undefined;
    }

    //expired time
    if ([NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (!data.expiredTime) return false;
        let expr: string = data.expiredTime;
        expr = calculateSubExpr(expr, args);
        if (data.unitExpiredTime === UNIT.MINUTE) {
            expr = `((${expr}) * 60000) + ${util.nextTime(args.data[0].startTime, args.timeframe)}`;
        }
        else if (data.unitExpiredTime === UNIT.CANDLE) {
            expr = `((${expr}) * 60000 * ${util.timeframeToNumberMinutes(args.timeframe)}) + ${util.nextTime(args.data[0].startTime, args.timeframe)}`;
        }
        data.expiredTime = calculate(expr, args);
        if (data.expiredTime === null) return false;
    }
    else {
        data.expiredTime = undefined;
    }

    // match TP, SL immediately
    if ([NODE_TYPE.BUY_MARKET, NODE_TYPE.BUY_LIMIT, NODE_TYPE.BUY_STOP_MARKET, NODE_TYPE.BUY_STOP_LIMIT].includes(data.type)) {
        if (data.entry <= data.sl) data.sl = data.entry;
        if (data.entry >= data.tp) data.tp = data.entry;
    }
    else if ([NODE_TYPE.SELL_MARKET, NODE_TYPE.SELL_LIMIT, NODE_TYPE.SELL_STOP_MARKET, NODE_TYPE.SELL_STOP_LIMIT].includes(data.type)) {
        if (data.entry >= data.sl) data.sl = data.entry;
        if (data.entry <= data.tp) data.tp = data.entry;
    }

    return true;
}

if (parentPort) {
    console.log('worker loaded');
    parentPort.on('message', async (msg: { type: string, value: any }) => {
        const t1 = new Date().getTime();
        if (msg.type === 'onCloseCandle') {
            const workerData: WorkerData = msg.value;
            const { broker, symbol, timeframe, data } = workerData;

            onCloseCandle(broker, symbol, timeframe, data);
        }
        else if (msg.type === 'setBotChildren') {
            botChildren = msg.value;
        }
        else if (msg.type === 'setBotIDs') {
            botIDs = msg.value;
        }

        const t2 = new Date().getTime();
        parentPort!.postMessage(t2 - t1);
    });
}
else {
    console.error(`Worker thread error.`)
    throw 'parentPort is null';
}

File: botfather_c++/src/Worker.cpp
#include "Worker.h"
#include "custom_indicator.h"
#include "util.h"
#include "Timer.h"
#include "expr.h"

static vector<string> orderTypes = {NODE_TYPE::BUY_MARKET, NODE_TYPE::BUY_LIMIT, NODE_TYPE::BUY_STOP_MARKET, NODE_TYPE::BUY_STOP_LIMIT, NODE_TYPE::SELL_MARKET, NODE_TYPE::SELL_LIMIT, NODE_TYPE::SELL_STOP_MARKET, NODE_TYPE::SELL_STOP_LIMIT};

Worker::Worker(shared_ptr<vector<shared_ptr<Bot>>> botList, string broker, string symbol, string timeframe, vector<double> open, vector<double> high, vector<double> low, vector<double> close, vector<double> volume, vector<long long> startTime)
    : botList(botList), broker(move(broker)), symbol(move(symbol)), timeframe(move(timeframe)), open(move(open)), high(move(high)), low(move(low)), close(move(close)), volume(move(volume)), startTime(move(startTime)), precision(5) {};

static int binarySearch(const vector<Symbol> &symbolList, const string &symbol)
{
    int left = 0;
    int right = symbolList.size() - 1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;
        if (symbolList[mid].symbolName == symbol)
        {
            return mid;
        }
        else if (symbolList[mid].symbolName < symbol)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }
    return -1;
}

void Worker::run()
{
    Timer timer(StringFormat("onCloseCandle %s %s %s", broker.c_str(), symbol.c_str(), timeframe.c_str()));
    for (const shared_ptr<Bot> &bot : *botList)
    {
        try
        {
            if (binarySearch(bot->symbolList, broker + ":" + symbol) == -1)
            {
                continue;
            }
            if (find(bot->timeframes.begin(), bot->timeframes.end(), timeframe) == bot->timeframes.end())
            {
                continue;
            }
            visited.clear();
            dfs_handleLogic(bot->route, bot->id);
        }
        catch (const exception &e)
        {
            LOGE("Error in bot %s: %s", bot->botName.c_str(), e.what());
            continue;
        }
    }
}

void Worker::dfs_handleLogic(Route &route, int botID)
{
    if (visited[route.id])
    {
        return;
    }
    visited[route.id] = true;
    if (handleLogic(route.data, botID))
    {
        for (Route &next : route.next)
        {
            dfs_handleLogic(next, botID);
        }
    }
}

string Worker::calculateSub(string &expr)
{
    return calculateSubExpr(expr, broker, symbol, timeframe, open.size(),
                            open.data(), high.data(), low.data(), close.data(), volume.data(),
                            startTime.data());
}

any Worker::calculate(string &expr)
{
    return calculateExpr(expr, broker, symbol, timeframe, open.size(),
                         open.data(), high.data(), low.data(), close.data(), volume.data(),
                         startTime.data());
}

bool Worker::adjustParam(NodeData &node)
{
    // stop
    if (node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::BUY_STOP_LIMIT)
    {
        if (node.stop.empty())
            return false;

        string expr = node.stop;
        expr = calculateSub(expr);
        if (node.unitStop == UNIT::PERCENT)
        {
            expr = StringFormat("close() * (100 + abs(%s)) / 100", expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate stop error. expr=%s", node.stop.c_str());
            return false;
        }

        node.stop = doubleToString(any_cast<double>(result), precision);
    }
    else if (node.type == NODE_TYPE::SELL_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (node.stop.empty())
            return false;

        string expr = node.stop;
        expr = calculateSub(expr);
        if (node.unitStop == UNIT::PERCENT)
        {
            expr = StringFormat("close() * (100 - abs(%s)) / 100", expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate stop error. expr=%s", node.stop.c_str());
            return false;
        }

        node.stop = doubleToString(any_cast<double>(result), precision);
    }
    else
    {
        node.stop = "";
    }

    // entry
    if (node.type == NODE_TYPE::BUY_LIMIT || node.type == NODE_TYPE::BUY_STOP_LIMIT)
    {
        if (node.entry.empty())
            return false;

        string expr = node.entry;
        expr = calculateSub(expr);
        if (node.unitEntry == UNIT::PERCENT)
        {
            expr = StringFormat("close() * (100 - abs(%s)) / 100", expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate entry error. expr=%s", node.entry.c_str());
            return false;
        }

        node.entry = doubleToString(any_cast<double>(result), precision);
    }
    else if (node.type == NODE_TYPE::SELL_LIMIT || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (node.entry.empty())
            return false;

        string expr = node.entry;
        expr = calculateSub(expr);
        if (node.unitEntry == UNIT::PERCENT)
        {
            expr = StringFormat("close() * (100 + abs(%s)) / 100", expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate entry error. expr=%s", node.entry.c_str());
            return false;
        }

        node.entry = doubleToString(any_cast<double>(result), precision);
    }
    else if (node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_MARKET)
    {
        node.entry = node.stop;
    }
    else if (node.type == NODE_TYPE::BUY_MARKET || node.type == NODE_TYPE::SELL_MARKET)
    {
        node.entry = doubleToString(close[0], precision);
    }

    double closePrice = close[0];
    double entry = stod(node.entry);
    double stop = node.stop.empty() ? 0 : stod(node.stop);
    // match entry immediately
    if (node.type == NODE_TYPE::BUY_LIMIT && closePrice <= entry)
    {
        node.entry = doubleToString(closePrice, precision);
    }
    else if (node.type == NODE_TYPE::BUY_STOP_LIMIT && closePrice <= entry && closePrice >= stop)
    {
        node.entry = doubleToString(closePrice, precision);
    }
    else if (node.type == NODE_TYPE::SELL_LIMIT && closePrice >= entry)
    {
        node.entry = doubleToString(closePrice, precision);
    }
    else if (node.type == NODE_TYPE::SELL_STOP_LIMIT && closePrice >= entry && closePrice <= stop)
    {
        node.entry = doubleToString(closePrice, precision);
    }

    // sl
    if (node.type == NODE_TYPE::BUY_MARKET || node.type == NODE_TYPE::BUY_LIMIT || node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::BUY_STOP_LIMIT)
    {
        if (node.sl.empty())
            return false;

        string expr = node.sl;
        expr = calculateSub(expr);
        if (node.unitSL == UNIT::PERCENT)
        {
            expr = StringFormat("(%s) * (100 - abs(%s)) / 100", node.entry.c_str(), expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate SL error. expr=%s", node.sl.c_str());
            return false;
        }

        node.sl = doubleToString(any_cast<double>(result), precision);
    }
    else if (node.type == NODE_TYPE::SELL_MARKET || node.type == NODE_TYPE::SELL_LIMIT || node.type == NODE_TYPE::SELL_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (node.sl.empty())
            return false;

        string expr = node.sl;
        expr = calculateSub(expr);
        if (node.unitSL == UNIT::PERCENT)
        {
            expr = StringFormat("(%s) * (100 + abs(%s)) / 100", node.entry.c_str(), expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate SL error. expr=%s", node.sl.c_str());
            return false;
        }

        node.sl = doubleToString(any_cast<double>(result), precision);
    }
    else
    {
        node.sl = "";
    }

    // tp
    if (node.type == NODE_TYPE::BUY_MARKET || node.type == NODE_TYPE::BUY_LIMIT || node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::BUY_STOP_LIMIT)
    {
        if (node.tp.empty())
            return false;

        string expr = node.tp;
        expr = calculateSub(expr);
        if (node.unitTP == UNIT::PERCENT)
        {
            expr = StringFormat("(%s) * (100 + abs(%s)) / 100", node.entry.c_str(), expr.c_str());
        }
        else if (node.unitTP == UNIT::RR)
        {
            expr = StringFormat("(%s + abs(%s - %s) * abs(%s))", node.entry.c_str(), node.entry.c_str(), node.sl.c_str(), expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate TP error. expr=%s", node.tp.c_str());
            return false;
        }

        node.tp = doubleToString(any_cast<double>(result), precision);
    }
    else if (node.type == NODE_TYPE::SELL_MARKET || node.type == NODE_TYPE::SELL_LIMIT || node.type == NODE_TYPE::SELL_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (node.tp.empty())
            return false;

        string expr = node.tp;
        expr = calculateSub(expr);
        if (node.unitTP == UNIT::PERCENT)
        {
            expr = StringFormat("(%s) * (100 - abs(%s)) / 100", node.entry.c_str(), expr.c_str());
        }
        else if (node.unitTP == UNIT::RR)
        {
            expr = StringFormat("(%s - abs(%s - %s) * abs(%s))", node.entry.c_str(), node.entry.c_str(), node.sl.c_str(), expr.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate TP error. expr=%s", node.tp.c_str());
            return false;
        }

        node.tp = doubleToString(any_cast<double>(result), precision);
    }
    else
    {
        node.tp = "";
    }

    // volume
    if (find(orderTypes.begin(), orderTypes.end(), node.type) != orderTypes.end())
    {
        if (node.volume.empty())
            return false;

        string expr = node.volume;
        expr = calculateSub(expr);
        if (node.unitVolume == UNIT::USD)
        {
            expr = StringFormat("(%s) / %s", expr.c_str(), node.entry.c_str());
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate volume error. expr=%s", node.volume.c_str());
            return false;
        }

        node.volume = doubleToString(any_cast<double>(result), precision);
    }
    else
    {
        node.volume = "";
    }

    // expired time
    if (node.type == NODE_TYPE::BUY_LIMIT || node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::BUY_STOP_LIMIT || node.type == NODE_TYPE::SELL_LIMIT || node.type == NODE_TYPE::SELL_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (node.expiredTime.empty() || node.expiredTime == "0")
            return false;

        string expr = node.expiredTime;
        expr = calculateSub(expr);
        if (node.unitExpiredTime == UNIT::MINUTE)
        {
            expr = StringFormat("((%s) * 60000) + %lld", expr.c_str(), nextTime(startTime[0], timeframe));
        }
        else if (node.unitExpiredTime == UNIT::CANDLE)
        {
            expr = StringFormat("((%s) * 60000 * %d) + %lld", expr.c_str(), timeframeToNumberMinutes(timeframe), nextTime(startTime[0], timeframe));
        }

        any result = calculate(expr);
        if (!result.has_value() || result.type() != typeid(double))
        {
            LOGE("Calculate expiredTime error. expr=%s", node.expiredTime.c_str());
            return false;
        }

        node.expiredTime = doubleToString(any_cast<double>(result), precision);
    }
    else
    {
        node.expiredTime = "";
    }

    // match TP, SL immediately
    entry = stod(node.entry);
    double sl = stod(node.sl);
    double tp = stod(node.tp);
    if (node.type == NODE_TYPE::BUY_MARKET || node.type == NODE_TYPE::BUY_LIMIT || node.type == NODE_TYPE::BUY_STOP_MARKET || node.type == NODE_TYPE::BUY_STOP_LIMIT)
    {
        if (entry <= sl)
            node.sl = node.entry;
        if (entry >= tp)
            node.tp = node.entry;
    }
    else if (node.type == NODE_TYPE::SELL_MARKET || node.type == NODE_TYPE::SELL_LIMIT || node.type == NODE_TYPE::SELL_STOP_MARKET || node.type == NODE_TYPE::SELL_STOP_LIMIT)
    {
        if (entry >= sl)
            node.sl = node.entry;
        if (entry <= tp)
            node.tp = node.entry;
    }

    return true;
}

bool Worker::handleLogic(NodeData &node, int botID)
{
    if (node.type == NODE_TYPE::START)
        return true;

    if (node.type == NODE_TYPE::EXPR)
    {
        any result = calculateExpr(node.value, broker, symbol, timeframe, open.size(),
                                   open.data(), high.data(), low.data(), close.data(), volume.data(),
                                   startTime.data());

        if (result.has_value())
        {
            if (result.type() == typeid(int))
            {
                return any_cast<int>(result) != 0;
            }
            else if (result.type() == typeid(double))
            {
                return any_cast<double>(result) != 0.0;
            }
            else if (result.type() == typeid(string))
            {
                return true;
            }
            else
            {
                LOGE("Unknown result type");
                return false;
            }
        }
        else
        {
            LOGE("No result");
            return false;
        }
    }
    if (node.type == NODE_TYPE::TELEGRAM)
    {
        string mess = calculateSub(node.value);
        LOGD("Send telegram message: %s %s %s %s", broker.c_str(), symbol.c_str(), timeframe.c_str(), mess.c_str());
        return true;
    }

    // new order
    if (!adjustParam(node))
    {
        return false;
    }

    if (find(orderTypes.begin(), orderTypes.end(), node.type) != orderTypes.end())
    {
        LOGI("New order - BotID: %d, Type: %s, Broker: %s, Symbol: %s, Timeframe: %s, Entry: %s, Stop: %s, TP: %s, SL: %s, Volume: %s, ExpiredTime: %s",
             botID, node.type.c_str(), broker.c_str(), symbol.c_str(), timeframe.c_str(),
             node.entry.c_str(), node.stop.c_str(), node.tp.c_str(), node.sl.c_str(),
             node.volume.c_str(), node.expiredTime.c_str());

        return true;
    }

    return false;
}